File: app/Support/Rotation/scripts/smooth_weighted_round_robin.lua
-- Smooth weighted round robin, mirroring App\Support\Rotation\SmoothWeightedRoundRobin.
--
-- KEYS[1]     state hash: destination id -> current weight
-- KEYS[2]     index set holding every state key used by this rotator
-- ARGV[1]     the unprefixed state key, recorded in the index set so the keys
--             can be deleted later without SCAN and without the client prefix
--             being applied twice
-- ARGV[2]     ttl in seconds
-- ARGV[3..n]  destination id, weight, destination id, weight, ...
--             ordered by destination id ascending; ties resolve to the first
--             entry, so that ordering is part of the behavioural contract.
--
-- The whole read, decide and write cycle runs inside one EVAL, and Redis runs
-- scripts to completion, so concurrent requests cannot lose an update.
--
-- Returns the winning destination id, or nil when no destinations were given.

local key = KEYS[1]
local index = KEYS[2]
local rawKey = ARGV[1]
local ttl = tonumber(ARGV[2])
local count = (#ARGV - 2) / 2

if count < 1 then
    return nil
end

local ids = {}
local weights = {}
local total = 0

for i = 1, count do
    ids[i] = ARGV[2 * i + 1]
    weights[i] = tonumber(ARGV[2 * i + 2])
    total = total + weights[i]
end

-- Read by explicit field name so a stale field left by an older fingerprint can
-- never be picked up. Missing fields come back as false, which becomes zero.
local stored = redis.call('HMGET', key, unpack(ids))
local current = {}
local sum = 0

for i = 1, count do
    local value = tonumber(stored[i])

    if value == nil then
        value = 0
    end

    current[i] = value
    sum = sum + value
end

-- The current weights always sum to exactly zero: each pass adds the total
-- across all destinations and subtracts the total from one. Any other sum means
-- the state was truncated or tampered with, so restart the cycle.
if sum ~= 0 then
    for i = 1, count do
        current[i] = 0
    end
end

local winner = 1

for i = 1, count do
    current[i] = current[i] + weights[i]

    if current[i] > current[winner] then
        winner = i
    end
end

current[winner] = current[winner] - total

local write = {}

for i = 1, count do
    write[#write + 1] = ids[i]
    write[#write + 1] = tostring(current[i])
end

redis.call('HSET', key, unpack(write))
redis.call('EXPIRE', key, ttl)

redis.call('SADD', index, rawKey)
redis.call('EXPIRE', index, ttl)

return ids[winner]

File: app/Support/Rotation/SmoothWeightedRoundRobin.php
<?php

namespace App\Support\Rotation;

/**
 * Smooth weighted round robin, as used by nginx's upstream balancer.
 *
 * This class is the normative definition of the algorithm. The Lua script in
 * scripts/smooth_weighted_round_robin.lua mirrors it so the Redis driver can run
 * the same selection atomically, and the two are held together by a shared
 * dataset of expected sequences.
 */
final class SmoothWeightedRoundRobin
{
    /**
     * Advance the cursor by one selection.
     *
     * @param  array<int, int>  $weights  Destination id => weight, ordered by destination id ascending.
     * @param  array<int, mixed>  $currentWeights  Destination id => stored current weight, from an untyped store.
     * @return array{destinationId: int|null, currentWeights: array<int, int>}
     */
    public function advance(array $weights, array $currentWeights): array
    {
        if ($weights === []) {
            return ['destinationId' => null, 'currentWeights' => []];
        }

        $total = array_sum($weights);
        $current = $this->normalize($weights, $currentWeights);

        $winner = null;

        foreach ($weights as $id => $weight) {
            $current[$id] += $weight;

            if ($winner === null || $current[$id] > $current[$winner]) {
                $winner = $id;
            }
        }

        $current[$winner] -= $total;

        return ['destinationId' => $winner, 'currentWeights' => $current];
    }

    /**
     * Drop unknown entries, zero fill missing ones and restart a broken cycle.
     *
     * The current weights always sum to exactly zero: each pass adds the total
     * across all destinations and subtracts the total from one. Any other sum
     * means the stored state was truncated or tampered with, so the cycle is
     * restarted rather than continued from a meaningless position.
     *
     * @param  array<int, int>  $weights
     * @param  array<int, mixed>  $currentWeights
     * @return array<int, int>
     */
    private function normalize(array $weights, array $currentWeights): array
    {
        $current = [];

        foreach ($weights as $id => $weight) {
            $stored = $currentWeights[$id] ?? 0;

            $current[$id] = is_int($stored) ? $stored : 0;
        }

        if (array_sum($current) !== 0) {
            return array_map(static fn (): int => 0, $current);
        }

        return $current;
    }
}
